src/submission/submission-service.ts

```diff
@@ -851,26 +859,110 @@ export namespace operations {
+    }
+  };
+
+  const findException = ({
+    exceptions,
+    record,
+    validationErrorFieldName,
+  }: {
+    exceptions: ExceptionRecords;
+    record: DataRecord;
+    validationErrorFieldName: string;
+  }): ProgramExceptionRecord | EntityExceptionRecord | undefined => {
+    if (isArrayOfEntityExceptionRecord(exceptions)) {
+      // entity exception
+      return exceptions.find(
+        exception =>
+          exception.requested_core_field === validationErrorFieldName &&
+          findExceptionByEntity({ exception, record }),
+      );
+    } else {
+      // program exception
+      return exceptions.find(exception => {
+        exception.requested_core_field === validationErrorFieldName;
+      });
+    }
+  };
+
+  const specimenKeyField: keyof SpecimenExceptionRecord = 'submitter_specimen_id';
+  const findExceptionByEntity = ({
+    exception,
+    record,
+  }: {
+    exception: EntityExceptionRecord;
+    record: DataRecord;
+  }) => {
+    // NB: keep type checks as string literals for current project TS version 3.9.5
+    if ('submitter_specimen_id' in exception) {
+      return exception[specimenKeyField] === record[specimenKeyField];
+    } else {
+      return false;
     }
-    // every other validation of SchemaValidationErrorTypes check for exception
-    else {
-      const validationErrorField = validationError.fieldName;
-      const exception = exceptions.find(
-        exception => exception.requested_core_field === validationErrorField,
+  };
+
+  const applyExceptions = async ({
+    programId,
+    entity,
+    record,
+    schemaValidationErrors,
+  }: {
+    programId: string;
+    entity: string;
+    record: dictionaryEntities.DataRecord;
+    schemaValidationErrors: dictionaryEntities.SchemaValidationError[];
+  }): Promise<dictionaryEntities.SchemaValidationError[]> => {
+    const t0 = performance.now();
+    // program exceptions and entity exceptions are mutually exclusive
+
+    // program level exceptions
+    const programExceptionResult = await programExceptionRepository.find(programId);
+    if (isProgramException(programExceptionResult)) {
+      return schemaValidationErrors.filter(
+        validationError =>
+          !isException({ exceptions: programExceptionResult.exceptions, validationError, record }),
       );
-      return exception?.requested_exception_value === record[validationErrorField];
     }
+
+    // entity level exceptions
+    const entityExceptionResult = await entityExceptionRepository.find(programId);
+    if (
+      isEntityException(entityExceptionResult) &&
+      (entity === EntityValues.followup || entity === EntityValues.specimen)
+    ) {
+      const entityExceptions = entityExceptionResult[entity];
+
+      return schemaValidationErrors.filter(
+        validationError => !isException({ exceptions: entityExceptions, validationError, record }),
+      );
+    }
+
+    const t1 = performance.now();
+    L.debug('apply exceptions time: ' + (t1 - t0));
```

Comment:
probably not something introduced ib this PR at all, but is L short for Logger? 😂

LOL.why();